File: src/frxx/proc/algs/bootstrapDPSD.py
from typing import TYPE_CHECKING, Tuple, TypeAlias

import numpy as np
from numpy.typing import NDArray


Complex64Array: TypeAlias = NDArray[np.complex64]
Complex128Array: TypeAlias = NDArray[np.complex128]
Float64Array: TypeAlias = NDArray[np.float64]

if TYPE_CHECKING:
    def _computeSingleSpectrum(
        VH: Complex64Array,
        VV: Complex64Array,
        w: Float64Array,
        M: int,
        NFT: int,
        B: float,
        r: float,
    ) -> tuple[Float64Array, Float64Array, Complex128Array]: ...

    def _computeMultipleSpectra(
        VH: Complex64Array,
        VV: Complex64Array,
        w: Float64Array,
        NK: int,
        M: int,
        NFT: int,
        B: float,
        r: float,
    ) -> tuple[Float64Array, Float64Array, Complex128Array]: ...

    def processRay_S_cpp(
        iqh: Complex64Array,
        iqv: Complex64Array,
        window: Float64Array,
        nBootstraps: float,
        K: int = 1,
        NFT: int = 1,
    ) -> tuple[Float64Array, Float64Array, Float64Array, Float64Array]: ...
else:
    from ._bootstrapDPSD import (
        _computeMultipleSpectra,
        _computeSingleSpectrum,
        processRay_S as processRay_S_cpp,
    )

import threading
_thread_lock = threading.Lock()

def processRay_S_torch(
    iqh: np.ndarray,          # (NK, M) complex64
    iqv: np.ndarray,          # (NK, M) complex64
    window: np.ndarray,       # (M,)
    nBootstraps: float,
    K: int = 1,
    NFT: int = 1
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    PyTorch reimplementation of processRay_S.

    Accepts numpy arrays, returns numpy arrays.
    Runs entirely in float32/complex64 on the target GPU device.
    """
    from frxx import BACKEND, Backend
    import torch

    # --- Resolve device from backend ---
    match BACKEND:
        case Backend.TORCH_CUDA:
            device = torch.device("cuda")
        case Backend.TORCH_MPS:
            device = torch.device("mps")
        case _:
            raise ValueError(f"Unsupported backend for PyTorch path: {BACKEND}")

    fdtype = torch.float32
    cdtype = torch.complex64

    NK, M = iqh.shape
    N = NK // K

    if NFT <= 1:
        NFT = M
    global allocation_count

    with _thread_lock:
        # --- Move inputs to device ---
        VH = torch.from_numpy(iqh).to(dtype=cdtype, device=device)
        VV = torch.from_numpy(iqv).to(dtype=cdtype, device=device)
        win = torch.from_numpy(window).to(dtype=fdtype, device=device)

        # --- Compute r and nr ---
        alpha = torch.mean(win ** 2).item()
        r = 0.5 - np.sqrt(alpha) * 0.5
        nr = int(round(M * r))

        # --- Compute CX_left, CX_right for each spectrum (NK,) ---
        VH_first = VH[:, 0]
        VH_last = VH[:, -1]
        VV_first = VV[:, 0]
        VV_last = VV[:, -1]

        guard = ((torch.abs(VH_first) < 1e-20) | (torch.abs(VH_last) < 1e-20) |
                (torch.abs(VV_first) < 1e-20) | (torch.abs(VV_last) < 1e-20))

        CX_left = 0.5 * (VH_first / VH_last + VV_first / VV_last)
        CX_right = 0.5 * (VH_last / VH_first + VV_last / VV_first)
        CX_left[guard] = 1.0 + 0j
        CX_right[guard] = 1.0 + 0j

        # --- Build extended signals XH, XV: (NK, Mx) ---
        left_h = VH[:, -nr:-1] * CX_left[:, None]
        right_h = VH[:, 1:nr] * CX_right[:, None]
        XH = torch.cat([left_h, VH, right_h], dim=1)

        left_v = VV[:, -nr:-1] * CX_left[:, None]
        right_v = VV[:, 1:nr] * CX_right[:, None]
        XV = torch.cat([left_v, VV, right_v], dim=1)

        Mx = XH.shape[1]

        allBootstraps = nBootstraps == 0 or (
            np.isinf(nBootstraps) and nBootstraps > 0
        )
        if allBootstraps:
            nBootstraps = Mx - M + 1
        elif (
            not np.isfinite(nBootstraps)
            or nBootstraps < 1
            or int(nBootstraps) != nBootstraps
        ):
            raise ValueError(
                "nBootstraps must be a positive integer, zero, or positive infinity"
            )
        else:
            nBootstraps = int(nBootstraps)

        # Free original tensors early
        del VH, VV

        # --- Compute R0 of original signals (from the middle of extended) ---
        middleStart = M - 1 if nr == 0 else nr - 1
        R0H = torch.mean(torch.abs(XH[:, middleStart:middleStart + M]) ** 2, dim=1)  # (NK,)
        R0V = torch.mean(torch.abs(XV[:, middleStart:middleStart + M]) ** 2, dim=1)  # (NK,)

        # --- Accumulate spectra across bootstraps ---
        SH_acc = torch.zeros((NK, NFT), dtype=fdtype, device=device)
        SV_acc = torch.zeros((NK, NFT), dtype=fdtype, device=device)
        SX_acc = torch.zeros((NK, NFT), dtype=cdtype, device=device)

        offsets = torch.arange(M, device=device)  # (M,)

        # --- Auto-calculate bootstrap_batch ---
        if device.type == "cuda":
            torch.cuda.synchronize()
            free_mem, _ = torch.cuda.mem_get_info(device)
        else:
            # MPS has no reliable free-memory query; fall back to a conservative estimate.
            # torch.mps.recommended_max_memory() exists but isn't a "free" query.
            import psutil
            free_mem = int(psutil.virtual_memory().available * 0.5)  # rough heuristic

        # Peak in-loop memory per unit of batch_size:
        # All of these coexist simultaneously at peak (before the del at loop end):
        #   indices_flat : NK * B * M * 8   (int64)
        #   blockH       : NK * B * M * 8   (complex64)
        #   blockV       : NK * B * M * 8   (complex64)
        #   fftH         : NK * B * NFT * 8 (complex64)
        #   fftV         : NK * B * NFT * 8 (complex64)
        #   temp from SX : NK * B * NFT * 8 (fftH * conj(fftV))
        #   boot_idx     : NK * B * 8       (int64, negligible)
        bytes_per_unit = NK * (2 * M * 8 + 3 * NFT * 8 + 8)

        # Target 75% of free memory, with 1.15x safety factor on the estimate
        bootstrap_batch = max(1, int((free_mem * 0.75) / (bytes_per_unit * 1.15)))
        bootstrap_batch = min(bootstrap_batch, nBootstraps)

        for batch_start in range(0, nBootstraps, bootstrap_batch):
            batch_size = min(bootstrap_batch, nBootstraps - batch_start)

            # Bootstrap start indices: (NK, batch_size)
            if allBootstraps:
                boot_idx = torch.arange(
                    batch_start, batch_start + batch_size, device=device
                ).expand(NK, -1)
            else:
                boot_idx = torch.randint(
                    0, Mx - M + 1, (NK, batch_size), device=device
                )

            # Gather indices: (NK, batch_size * M)
            indices = boot_idx.unsqueeze(2) + offsets.unsqueeze(0).unsqueeze(0)
            indices_flat = indices.reshape(NK, batch_size * M)

            def gather_complex(x, dim, index):
                if x.is_complex() and x.device.type == "mps":
                    real = torch.gather(x.real, dim, index)
                    imag = torch.gather(x.imag, dim, index)
                    return torch.complex(real, imag)
                else:
                    return torch.gather(x, dim, index)


            blockH = gather_complex(XH, 1, indices_flat).reshape(NK, batch_size, M)
            blockV = gather_complex(XV, 1, indices_flat).reshape(NK, batch_size, M)

            del indices, indices_flat, boot_idx  # free before FFT

            # Compute R0 of each bootstrap block: (NK, batch_size)
            r0h_block = torch.mean(torch.abs(blockH) ** 2, dim=2).clamp(min=1e-20)
            r0v_block = torch.mean(torch.abs(blockV) ** 2, dim=2).clamp(min=1e-20)

            # Scale: (NK, batch_size)
            scaleH = torch.sqrt(R0H[:, None] / r0h_block)
            scaleV = torch.sqrt(R0V[:, None] / r0v_block)

            # Apply scale and window: (NK, batch_size, M)
            blockH = blockH * (scaleH.unsqueeze(2) * win.unsqueeze(0).unsqueeze(0))
            blockV = blockV * (scaleV.unsqueeze(2) * win.unsqueeze(0).unsqueeze(0))

            # FFT along last dim: (NK, batch_size, NFT)
            fftH = torch.fft.fft(blockH, n=NFT, dim=2)
            fftV = torch.fft.fft(blockV, n=NFT, dim=2)

            # Accumulate power and cross-spectrum, summing over batch dim
            SH_acc += (fftH.real ** 2 + fftH.imag ** 2).sum(dim=1)
            SV_acc += (fftV.real ** 2 + fftV.imag ** 2).sum(dim=1)
            SX_acc += (fftH * fftV.conj()).sum(dim=1)

            del blockH, blockV, fftH, fftV  # if still in scope

        # --- Normalize ---
        norm = M * alpha * nBootstraps
        SH_acc /= norm
        SV_acc /= norm
        SX_acc /= norm

        # --- fftshift ---
        SH_acc = torch.fft.fftshift(SH_acc, dim=1)
        SV_acc = torch.fft.fftshift(SV_acc, dim=1)
        SX_acc = torch.fft.fftshift(SX_acc, dim=1)

        # --- Average over K looks ---
        SH_avg = SH_acc.reshape(N, K, NFT).mean(dim=1)  # (N, NFT)
        SV_avg = SV_acc.reshape(N, K, NFT).mean(dim=1)
        SX_avg = SX_acc.reshape(N, K, NFT).mean(dim=1)

        # --- Bias correction (ZDR, RHOHV) ---
        if K == 1:
            beta = (1 - r) ** (-3.3) - 2 * ((1 - r) ** 1.1)
        else:
            beta = (1 - r) ** (-4.5) - (1 - r) ** (-2.1)

        td = SH_avg / SV_avg.clamp(min=1e-20)

        denom = torch.sqrt((SH_avg * SV_avg).clamp(min=0)).clamp(min=1e-20)
        tr = torch.abs(SX_avg) / denom

        trsquared = (tr ** 2).clamp(min=1e-20)

        sZDR = td * (1 - (1 / (beta * K)) * (1 - trsquared))
        sRHOHV = tr * (1 - (1 / (beta * K)) * ((1 - trsquared) ** 2 / (4 * trsquared)))

        # --- NaN / clamp handling ---
        bad = (SV_avg < 1e-20) | (denom < 1e-20)

        PSDH = SH_avg.clone()
        PSDV = SV_avg.clone()
        PSDH[PSDH < 0] = float('nan')
        PSDV[PSDV < 0] = float('nan')
        sZDR[bad] = float('nan')
        sZDR[sZDR < 0] = float('nan')
        sRHOHV[bad] = float('nan')
        sRHOHV[sRHOHV < 0] = 0.0

        # --- Convert to dB, move to CPU, return as numpy ---
        psdh_db = (10 * torch.log10(PSDH)).cpu().numpy()
        psdv_db = (10 * torch.log10(PSDV)).cpu().numpy()
        szdr_db = (10 * torch.log10(sZDR)).cpu().numpy()
        rhohv = sRHOHV.cpu().numpy()

        # Explicitly free ALL GPU tensors before releasing the lock
        del XH, XV, SH_acc, SV_acc, SX_acc
        del SH_avg, SV_avg, SX_avg
        del PSDH, PSDV, sZDR, sRHOHV
        del td, tr, trsquared, denom, bad
        del R0H, R0V, win
        #torch.cuda.empty_cache()

    return psdh_db, psdv_db, szdr_db, rhohv

def processRay_S(
    iqh: np.ndarray,          # (NK, M) complex64
    iqv: np.ndarray,          # (NK, M) complex64
    window: np.ndarray,       # (M,)
    nBootstraps: float,
    K: int = 1,
    NFT: int = 1
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    from frxx import BACKEND, Backend
    if BACKEND == Backend.NUMBA:
        return processRay_S_cpp(iqh, iqv, window, nBootstraps, K, NFT)
    else:
        return processRay_S_torch(iqh, iqv, window, nBootstraps, K, NFT)

File: src/cpp/frxx/proc/algs/bootstrapDPSD.cpp
#include <frxx/proc/algs/bootstrapDPSD.hpp>

#include <algorithm>
#include <cmath>
#include <complex>
#include <cstddef>
#include <limits>
#include <optional>
#include <random>
#include <stdexcept>
#include <utility>
#include <vector>

#include <pocketfft/pocketfft.hpp>

#include <frxx/utils/workerPool.hpp>

namespace frxx::proc::algs::bootstrap_dpsd {

namespace {

using frxx::utils::i64;
using Complex64Array2D = frxx::eigen::Array2D<std::complex<float>>;

struct BootstrapSetup {
    i64 B;
    bool allBootstraps;
};

BootstrapSetup resolve_bootstraps(double nBootstraps, i64 nPossibleBootstraps) {
    if (nBootstraps == 0.0 ||
        (std::isinf(nBootstraps) && nBootstraps > 0.0)) {
        return {nPossibleBootstraps, true};
    }
    if (!std::isfinite(nBootstraps) || nBootstraps < 1.0 ||
        std::floor(nBootstraps) != nBootstraps ||
        nBootstraps > static_cast<double>(std::numeric_limits<i64>::max())) {
        throw std::invalid_argument(
            "nBootstraps must be a positive integer, zero, or positive infinity");
    }
    return {static_cast<i64>(nBootstraps), false};
}

void fft_rows(Complex64Array2D& W, i64 B, i64 NFT) {
    const pocketfft::shape_t shape{
        static_cast<std::size_t>(B), static_cast<std::size_t>(NFT)};
    const pocketfft::stride_t stride{
        static_cast<std::ptrdiff_t>(NFT * sizeof(std::complex<float>)),
        static_cast<std::ptrdiff_t>(sizeof(std::complex<float>))};
    const pocketfft::shape_t axes{1};
    pocketfft::c2c(
        shape, stride, stride, axes, true,
        W.data(), W.data(), 1.0F, 1);
}

void require_single_spectrum_inputs(
    frxx::eigen::ConstArray1DRef<std::complex<float>> VH,
    frxx::eigen::ConstArray1DRef<std::complex<float>> VV,
    frxx::eigen::ConstArray1DRef<double> w,
    i64 M,
    i64 NFT
) {
    if (M <= 0) {
        throw std::invalid_argument("M must be positive");
    }
    if (NFT <= 0) {
        throw std::invalid_argument("NFT must be positive");
    }
    if (VH.size() != M || VV.size() != M || w.size() != M) {
        throw std::invalid_argument("VH, VV, and w must have length M");
    }
}

}  // namespace

SingleSpectrumResult computeSingleSpectrum(
    frxx::eigen::ConstArray1DRef<std::complex<float>> VH,
    frxx::eigen::ConstArray1DRef<std::complex<float>> VV,
    frxx::eigen::ConstArray1DRef<double> w,
    i64 M,
    i64 NFT,
    double nBootstraps,
    double r
) {
    require_single_spectrum_inputs(VH, VV, w, M, NFT);

    // Guard CX
    std::complex<double> CX_left;
    std::complex<double> CX_right;
    if (std::abs(VH(M - 1)) < 1e-30 || std::abs(VH(0)) < 1e-30 ||
        std::abs(VV(M - 1)) < 1e-30 || std::abs(VV(0)) < 1e-30) {
        CX_left = std::complex<float>(1.0F);
        CX_right = std::complex<float>(1.0F);
    } else {
        CX_left = 0.5 * (
            static_cast<std::complex<double>>(VH(0)) /
                static_cast<std::complex<double>>(VH(M - 1)) +
            static_cast<std::complex<double>>(VV(0)) /
                static_cast<std::complex<double>>(VV(M - 1)));
        CX_right = 0.5 * (
            static_cast<std::complex<double>>(VH(M - 1)) /
                static_cast<std::complex<double>>(VH(0)) +
            static_cast<std::complex<double>>(VV(M - 1)) /
                static_cast<std::complex<double>>(VV(0)));
    }

    const i64 nr = static_cast<i64>(std::nearbyint(static_cast<double>(M) * r));
    const i64 negnr = -nr;
    if (nr < 0 || nr > M) {
        throw std::invalid_argument("round(M * r) must be between zero and M");
    }

    const i64 leftStart = negnr == 0 ? 0 : M + negnr;
    const i64 leftSize = M - 1 - leftStart;
    const i64 rightSize = std::max<i64>(0, nr - 1);
    const i64 Mx = leftSize + M + rightSize;
    frxx::eigen::Array1D<std::complex<double>> XH(Mx);
    frxx::eigen::Array1D<std::complex<double>> XV(Mx);

    i64 x = 0;
    for (i64 j = leftStart; j < M - 1; ++j, ++x) {
        XH(x) = static_cast<std::complex<double>>(VH(j)) * CX_left;
        XV(x) = static_cast<std::complex<double>>(VV(j)) * CX_left;
    }
    for (i64 j = 0; j < M; ++j, ++x) {
        XH(x) = static_cast<std::complex<double>>(VH(j));
        XV(x) = static_cast<std::complex<double>>(VV(j));
    }
    for (i64 j = 1; j < nr; ++j, ++x) {
        XH(x) = static_cast<std::complex<double>>(VH(j)) * CX_right;
        XV(x) = static_cast<std::complex<double>>(VV(j)) * CX_right;
    }

    // R0 with no temporaries
    double accH = 0.0;
    double accV = 0.0;
    for (i64 j = 0; j < M; ++j) {
        const auto vh = VH(j);
        const auto vv = VV(j);
        accH += static_cast<double>(
            vh.real() * vh.real() + vh.imag() * vh.imag());
        accV += static_cast<double>(
            vv.real() * vv.real() + vv.imag() * vv.imag());
    }
    const double R0H = accH / static_cast<double>(M);
    const double R0V = accV / static_cast<double>(M);

    const i64 nPossibleBootstraps = Mx - M + 1;
    const auto bootstrapSetup = resolve_bootstraps(
        nBootstraps, nPossibleBootstraps);
    const i64 B = bootstrapSetup.B;

    // Fused: bootstrap + R0 + rescale + window in one parallel loop
    // The original implementation used range, rather than prange, here.
    //
    // WH and WV include the zero padding required by NFT. pocketfft transforms
    // every bootstrap row in-place, avoiding separate zH and zV allocations.
    Complex64Array2D WH(B, NFT);
    Complex64Array2D WV(B, NFT);
    WH.setZero();
    WV.setZero();

    std::optional<std::mt19937_64> randomEngine;
    if (!bootstrapSetup.allBootstraps) {
        randomEngine.emplace(std::random_device{}());
    }
    std::uniform_int_distribution<i64> randomBootIdx(
        0, nPossibleBootstraps - 1);

    for (i64 i = 0; i < B; ++i) {
        const i64 boot_idx = bootstrapSetup.allBootstraps
            ? i
            : randomBootIdx(*randomEngine);

        // Pass 1: extract block and accumulate |x|^2 for R0
        accH = 0.0;
        accV = 0.0;
        for (i64 j = 0; j < M; ++j) {
            const auto vh = XH(boot_idx + j);
            const auto vv = XV(boot_idx + j);
            accH += vh.real() * vh.real() + vh.imag() * vh.imag();
            accV += vv.real() * vv.real() + vv.imag() * vv.imag();
        }

        // Guard bootstrap scale
        if (accH < 1e-30) {
            accH = 1e-30;
        }
        if (accV < 1e-30) {
            accV = 1e-30;
        }
        // Pass 2: rescale + window in-place (row is hot in L1)
        const double scaleH = std::sqrt(R0H * M / accH);
        const double scaleV = std::sqrt(R0V * M / accV);
        for (i64 j = 0; j < std::min(M, NFT); ++j) {
            // Match the original complex64 WH/WV storage before the FFT.
            const auto wh = static_cast<std::complex<float>>(XH(boot_idx + j));
            const auto wv = static_cast<std::complex<float>>(XV(boot_idx + j));
            WH(i, j) = static_cast<std::complex<float>>(
                static_cast<std::complex<double>>(wh) * (scaleH * w(j)));
            WV(i, j) = static_cast<std::complex<float>>(
                static_cast<std::complex<double>>(wv) * (scaleV * w(j)));
        }
    }

    // FFT
    fft_rows(WH, B, NFT);
    fft_rows(WV, B, NFT);

    // Spectral averages — parallelize over frequency bins
    // The original implementation used range, rather than prange, here too.
    double alpha = 0.0;
    for (i64 j = 0; j < M; ++j) {
        alpha += w(j) * w(j);
    }
    alpha /= static_cast<double>(M);
    const double norm = static_cast<double>(M) * alpha * B;

    SingleSpectrumResult result{
        frxx::eigen::Array1D<double>(NFT),
        frxx::eigen::Array1D<double>(NFT),
        frxx::eigen::Array1D<std::complex<double>>(NFT)};

    const i64 half = NFT / 2;
    for (i64 j = 0; j < NFT; ++j) {
        double sh = 0.0;
        double sv = 0.0;
        std::complex<double> sx{0.0, 0.0};
        for (i64 i = 0; i < B; ++i) {
            const auto zh = static_cast<std::complex<double>>(WH(i, j));
            const auto zv = static_cast<std::complex<double>>(WV(i, j));
            sh += zh.real() * zh.real() + zh.imag() * zh.imag();
            sv += zv.real() * zv.real() + zv.imag() * zv.imag();
            sx += zh * std::conj(zv);
        }
        const i64 k = (j + half) % NFT;
        result.SHi(k) = sh / norm;
        result.SVi(k) = sv / norm;
        result.SXi(k) = sx / norm;
    }

    return result;
}

MultipleSpectraResult computeMultipleSpectra(
    frxx::eigen::ConstArray2DRef<std::complex<float>> VH,
    frxx::eigen::ConstArray2DRef<std::complex<float>> VV,
    frxx::eigen::ConstArray1DRef<double> w,
    i64 NK,
    i64 M,
    i64 NFT,
    double nBootstraps,
    double r
) {
    if (VH.rows() != NK || VV.rows() != NK ||
        VH.cols() != M || VV.cols() != M) {
        throw std::invalid_argument("VH and VV must have shape (NK, M)");
    }
    if (w.size() != M) {
        throw std::invalid_argument("w must have length M");
    }

    MultipleSpectraResult result{
        frxx::eigen::Array2D<double>(NK, NFT),
        frxx::eigen::Array2D<double>(NK, NFT),
        frxx::eigen::Array2D<std::complex<double>>(NK, NFT)};

    frxx::utils::WorkerPool pool;
    pool.pfor(0, NK, [&](i64 i) {
        auto VH_view = VH.row(i).transpose();
        auto VV_view = VV.row(i).transpose();
        const auto spectrum = computeSingleSpectrum(
            VH_view, VV_view, w,
            M, NFT, nBootstraps, r);
        result.SH.row(i) = spectrum.SHi.transpose();
        result.SV.row(i) = spectrum.SVi.transpose();
        result.SX.row(i) = spectrum.SXi.transpose();
    });

    return result;
}

ProcessRayResult processRay_S(
    frxx::eigen::ConstArray2DRef<std::complex<float>> iqh,
    frxx::eigen::ConstArray2DRef<std::complex<float>> iqv,
    frxx::eigen::ConstArray1DRef<double> window,
    double nBootstraps,
    i64 K,
    i64 NFT
) {
    if (iqh.rows() != iqv.rows() || iqh.cols() != iqv.cols()) {
        throw std::invalid_argument("iqh and iqv must have equal shapes");
    }
    if (K <= 0) {
        throw std::invalid_argument("K must be positive");
    }
    if (iqh.cols() <= 0 || window.size() != iqh.cols()) {
        throw std::invalid_argument("window must have one value per IQ pulse");
    }

    double meanWindowPower = 0.0;
    for (i64 j = 0; j < window.size(); ++j) {
        meanWindowPower += window(j) * window(j);
    }
    meanWindowPower /= static_cast<double>(window.size());
    const double r = 0.5 - std::sqrt(meanWindowPower) * 0.5;

    const i64 NK = iqh.rows();
    const i64 M = iqh.cols();

    const i64 N = NK / K;

    if (NFT <= 1) {
        NFT = M;
    }

    auto spectra = computeMultipleSpectra(
        iqh, iqv, window,
        NK, M, NFT, nBootstraps, r);
    auto& SH = spectra.SH;
    auto& SV = spectra.SV;
    auto& SX = spectra.SX;

    frxx::eigen::Array2D<double> tsh(N, NFT);
    frxx::eigen::Array2D<double> tsv(N, NFT);
    frxx::eigen::Array2D<std::complex<double>> tsx(N, NFT);
    frxx::eigen::Array2D<double> td(N, NFT);
    frxx::eigen::Array2D<double> tr(N, NFT);

    for (i64 i = 0; i < N; ++i) {
        for (i64 j = 0; j < NFT; ++j) {
            double sh = 0.0;
            double sv = 0.0;
            std::complex<double> sx{0.0, 0.0};
            for (i64 k = 0; k < K; ++k) {
                sh += SH(i * K + k, j);
                sv += SV(i * K + k, j);
                sx += SX(i * K + k, j);
            }
            tsh(i, j) = sh / K;
            tsv(i, j) = sv / K;
            tsx(i, j) = sx / static_cast<double>(K);
            if (tsv(i, j) < 1e-30) {
                td(i, j) = std::numeric_limits<double>::quiet_NaN();
                tr(i, j) = std::numeric_limits<double>::quiet_NaN();
            } else {
                td(i, j) = tsh(i, j) / tsv(i, j);
                const double denom = std::sqrt(tsh(i, j) * tsv(i, j));
                if (denom < 1e-30) {
                    tr(i, j) = std::numeric_limits<double>::quiet_NaN();
                } else {
                    tr(i, j) = std::abs(tsx(i, j)) / denom;
                }
            }
        }
    }

    double beta;
    if (K == 1) {
        beta = std::pow(1 - r, -3.3) - 2 * std::pow(1 - r, 1.1);
    } else {
        beta = std::pow(1 - r, -4.5) - std::pow(1 - r, -2.1);
    }

    auto& PSDH = tsh;
    auto& PSDV = tsv;
    // COV = tsx

    frxx::eigen::Array2D<double> sZDR(N, NFT);
    frxx::eigen::Array2D<double> sRHOHV(N, NFT);
    for (i64 i = 0; i < N; ++i) {
        for (i64 j = 0; j < NFT; ++j) {
            double trsquared = std::pow(tr(i, j), 2);
            if (trsquared < 1e-30) {
                trsquared = 1e-30;
            }

            sZDR(i, j) = td(i, j) *
                (1 - (1 / (beta * K) * (1 - trsquared)));
            sRHOHV(i, j) = tr(i, j) *
                (1 - (1 / (beta * K) *
                    (std::pow(1 - trsquared, 2) / (4 * trsquared))));

            if (PSDH(i, j) < 0) {
                PSDH(i, j) = std::numeric_limits<double>::quiet_NaN();
            }
            if (PSDV(i, j) < 0) {
                PSDV(i, j) = std::numeric_limits<double>::quiet_NaN();
            }
            if (sZDR(i, j) < 0) {
                sZDR(i, j) = std::numeric_limits<double>::quiet_NaN();
            }
            if (sRHOHV(i, j) < 0) {
                sRHOHV(i, j) = 0;
            }

            PSDH(i, j) = 10 * std::log10(PSDH(i, j));
            PSDV(i, j) = 10 * std::log10(PSDV(i, j));
            sZDR(i, j) = 10 * std::log10(sZDR(i, j));
        }
    }

    // return 10*log10(PSDH), 10*log10(PSDV), COV, 10*log10(sZDR), sRHOHV
    return {
        std::move(PSDH), std::move(PSDV),
        std::move(sZDR), std::move(sRHOHV)};
}

}  // namespace frxx::proc::algs::bootstrap_dpsd
